proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1269 @@
+   * Visible universal quantification: ``forall a -> b``.
+
+   We will call them **types-in-terms**.
+
+   Grammatically, their constituents are terms, not types::
+
+                   proposed grammar:                      as opposed to:
+         ┌────────────────────────────────────┬───────────────────────────────────────┐
+         │                                    │                                       │
+         │  exp ::=                           │    exp ::=                            │
+         │      | exp₀ '->' exp₁              │        | type₀ '->' type₁             │
+         │      | exp₀ '=>' exp₁              │        | type₀ '=>' type₁             │
+         │      | 'forall' tv_bndrs '.'  exp  │        | 'forall' tv_bndrs '.'  type  │
+         │      | 'forall' tv_bndrs '->' exp  │        | 'forall' tv_bndrs '->' type  │
+         │                                    │                                       │
+         └────────────────────────────────────┴───────────────────────────────────────┘
+
+   This is a necessity to avoid parsing conflicts, with the following
+   consequences:
+
+   1. The ``'`` symbol signifies Template Haskell name quotation rather than ``DataKinds`` promotion.
+   2. The ``*`` symbol is treated as an infix operator regardless of ``-XStarIsType``.
+   3. Built-in syntax for tuples and lists is interpreted as in terms.
+      That is, ``[a]`` is a singleton list rather than the type of a list,
+      and ``(a, b)`` is a pair rather than the type of a pair.
+
+2. Make ``forall`` a keyword at the term level. Not guarded by any extension
+   (same motivation as `#193 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0193-forall-keyword.rst>`_).
+   This implies ``forall`` is no longer a valid identifier.
+
+3. Introduce a new extension, ``ListTupleTypeSyntax``, on by default,
+   which enables:
+
+   * ``[]`` as the list type constructor
+   * ``()`` as the unit type
+   * ``[a]`` the syntax of a list type
+   * ``(,)`` as the pair type constructor
+   * ``(a, b)`` as the syntax of a pair type
+
+   When the extension is on, these constructs retain their Haskell 2010
+   meaning, which depends on whether we are in a type-level or term-level
+   context.
+
+   When the extension is off, all of the above are interpreted as in terms:
+
+   * ``[]`` is always an empty list
+   * ``()`` is always the unit data constructor
+   * ``[a]`` is always a singleton list (not the list type)
+   * ``(,)`` is always the pair data constructor (not the type constructor)
+   * ``(a, b)`` is always a pair (not the type of a pair)
+
+   Likewise for tuples of higher arities.
+
+   Export the following synonym from the ``Data.List`` module::
+
+     type List = []
+
+   Export the following synonyms from the ``Data.Tuple`` module::
+
+     type Unit = ()
+     type TupleOf2 = (,)
+     type TupleOf3 = (,,)
+     type TupleOf4 = (,,,)
+     ... -- up to the maximum tuple arity
```

Comment:
I am sure this was discussed elsewhere, but why not `type Tuple 2 = (,)`, i.e. a type family taking a `Nat`?

Reply:
What I’d actually like to do is to make `Tuple` a variadic data family, as I explain in https://github.com/ghc-proposals/ghc-proposals/pull/281#issuecomment-860163326

But the problem is that data families cannot be promoted. Once this restriction is lifted, this would certainly be an attractive option.

Reply:
@goldfirere How hard would it be to allow promotion of data families? Actually, let’s be less ambitious: how hard would it be to allow promotion of a single, hard-coded data family (i.e. `Tuple`)?

If you deem it possible, I will modify the proposal to introduce `Tuple` instead of numerous `TupleOfN` synonyms.